[L2] Add rSeries compatibility for L2 code

BenjaminLudwigSAP opened the pull request:
This is a draft PR for early review. I might "recut" the commits on Friday to make them clearer. But the content won't change (apart from tests).
I strongly recommend reviewing commit by commit.

BenjaminLudwigSAP committed:
Add BigIPRestClient support for rSeries F5OS-A API

velp commented on octavia_f5/controller/worker/l2_sync_manager.py:
```diff
@@ -27,8 +27,7 @@
 from octavia_f5.common import constants
-from octavia_f5.controller.worker.flows import f5_flows
-from octavia_f5.controller.worker.tasks import f5_tasks
+from octavia_f5.controller.worker.flows import f5_flows, f5_flows_rseries
```
If you have a specific name for a module that is working with R-series devices, you should have also specific name for old one (something like `f5_flows_bigip` instead of `f5_flows`)

velp commented on octavia_f5/controller/worker/flows/f5_flows.py:
```diff
@@ -18,11 +18,16 @@
 class F5Flows(object):
+
+    def __init__(self, tasks=f5_tasks):
+        self.tasks = tasks.F5Tasks()
```
Why did you decide to do it like this and put all task classes into one big class even for the old module? Why cannot we choose/pass the module that we need and keep naming for the tasks the same in both of them? In this case you do not need to change old  module (put everything to one big class) much and the code is much simplier for review.

BenjaminLudwigSAP replied:
I know, and I find this ugly as well. I'm trying to implement a better solution.
The reason for this decision is that while there are many tasks that differ between iSeries and rSeries, many others don't. And since I want the interface for F5 tasks to be the same, no matter for what backend they're being used, I tried to utilize inheritance - that is, to have `f5_tasks_rseries` inherit from `f5_tasks` and override only the task classes that are different for rSeries.
I tihnk there is no perfect solution, but I'm trying to change this to one that produces far less diff. I'll keep you updated.

velp reviewed (changes requested):
The PR has lots of unrelated to the task changes. Putting functions in alphabetical order, renaming existing variables where it doesn't matter. All of this makes review much more difficult, as does moving all classes into one big class when it could have been avoided.

On the other hand, even the basic logic is not ready now, because there are no tests and the tests that are there now clearly have a logical error:
```
./octavia_f5/controller/worker/tasks/f5_tasks_rseries.py:80:69: F821 undefined name 'vlan_id'
                LOG.warning(f"EnsureVLAN revert: Not deleting VLAN {vlan_id}, since it existed before "
                                                                    ^
1     F821 undefined name 'vlan_id'
```

P.S. Please, try to make only necessary changes in pull requests and keep them as simple as possible. This speeds up the review process and reduces the number of errors. Reviewing a change in which 1044 lines were added and 647 were removed seems impossible, especially if most of the changes were made by simply shifting lines.

velp commented on octavia_f5/restclient/bigip/bigip_restclient.py:
```diff
@@ -117,6 +119,7 @@ def update_status(self):
         return self._active
 
     @_metric_get_exceptions.count_exceptions()
+    @injectF5OSAHeaders
```
It is not obvious that the parameters change on the fly. It is better to make a method inside the class for adding headers and call it in the necessary places. One line is also added, like decorators, but this is much clearer and more obvious.

velp commented on octavia_f5/utils/decorators.py:
```diff
@@ -38,21 +38,59 @@ def wrapper(*args, **kwargs):
+        self.exception_class = F5osaException
+
+
+def injectF5OSAHeaders(f):
+    def wrapper(self, *args, **kwargs):
+        if self.f5os_a:
+            headers = kwargs.get('headers', {})
+            headers['Content-Type'] = 'application/yang-data+json'
+            headers['Accept'] = 'application/yang-data+json'
+            kwargs['headers'] = headers
+        return f(self, *args, **kwargs)
+    return wrapper
```
As I said above better to use a method and call it where it needed; the decorator is not clear for understanding when it's patching arguments on the way.

velp commented on octavia_f5/controller/worker/l2_sync_manager.py:
```diff
@@ -209,13 +212,14 @@ def ensure_l2_flow(self, selfips: [network_models.Port], network_id: str, device
+                data=ensure_l2_flow_data)] = self._bigips
 
-        # run VCMP l2 flow for all VCMPs in parallel
+        # run VCMP l2 flow for all VCMP hosts in parallel
```
Is it really related to this PR?

BenjaminLudwigSAP replied:
Yes, this PR is about vCMP hosts. This comment is about vCMP hosts as well. I just thought it could be clearer, so I made this little change.

velp reviewed (changes requested):
I want to add a more or less substantial review for those parts of the code that were changed as needed. The most important part of this change is in file octavia_f5/controller/worker/tasks/f5_tasks_rseries.py, but since there are no tests covering this functionality, I can only roughly check its logic syntactically.

Without adding unit tests that check the queries and show what logic is expected, I will not be able to approve this pull request.

BenjaminLudwigSAP committed:
Add rSeries flow tests

BenjaminLudwigSAP committed:
Ignore failed VLAN deletion due to bug - VLAN can safely be reused

This has been agreed on with @m-kratochvil and @simonkowallik

BenjaminLudwigSAP committed:
README: Clear up use of naming conventions in context of rSeries

BenjaminLudwigSAP committed:
Implement vCMP host tasks for rSeries F5OS-A API

- Add f5_tasks_rseries and f5_flows_rseries
  and rename f5_tasks to f5_tasks_iseries and f5_flows to f5_flows_iseries.
  f5_tasks_rseries inherits from f5_tasks_iseries (by importing and reexposing all tasks), but redefines tasks that are different for rSeries F5OS-A API. At the moment this only affects tasks operating on the vCMP host, not the guest: EnsureVLAN, GetExistingVLAN, EnsureVLANInterface, EnsureGuestVLAN, GetVCMPGuests, RemoveGuestVLAN, RemoveVLANInterface, RemoveVLANIfNotOwnedByGuest
  f5_flows_rseries simply inherits from f5_flows_iseries but configures it to use f5_tasks_rseries instead of f5_tasks_iseries.

- Patch L2SyncManager to use two different F5Flows instances - one for host tasks (which might be iSeries or rSeries) and one for guest tasks (which are always iSeries)

- Add RemoveVLANInterface task to vCMP host removal flow. On iSeries it's a NOP, but on rSeries it's needed.

- Since the helper functions defined in f5_tasks.py need to be consumable, irrespective of whether iSeries or rSeries tasks are used, move those helper functions to driver_utils.py
  - selfip_for_subnet_exists
  - get_subnet_route_name

- Add some clarifying comments to existing tasks where they help

- Remove return statement for tasks that that don't have default_provides set and/or whose outputs are never used
  - EnsureGuestVLAN: Output is never used. (Don't confuse `device_guest` with `device_guests` (with an s at the end), which is the output of GetVCMPGuests and is being used by other tasks!)

- Replace RaisesIControlRestError by RaisesApiError, which is used by both RaisesIControlRestError and RaisesF5osaError, but with different exceptions

velp reviewed (approved):
Please use Squash and Merge in the end.

m-kratochvil reviewed (approved):
Looks good, thanks for the major effort! The minor outstanding issue with vlan removal is understood and kept on separate track.
